bpf: nodeport: integrate Ingress RevSNAT and RevDNAT paths

julianwiedmann opened the pull request:
The Ingress RevSNAT and RevDNAT handling is currently split into two separate tail-calls. Which means that if the RevSNAT code performs a CT lookup and receives trace information, then we can't easily pass the corresponding `trace_ctx` to the `TO_OVERLAY` trace event in the RevDNAT code.

In practice this means that when in direct-routing mode, the trace events for EgressGW reply traffic don't benefit from CT-driven aggregation.

This PR refactors the relevant code paths, and turns the RevDNAT path into a conditional tail-call. On newer kernels it gets inlined into `tail_nodeport_nat_ingress_ipv4()`, and we can pass through the `trace_ctx` struct all the way to the `send_trace_notify()`.

julianwiedmann committed:
bpf: nodeport: shuffle IPv6 Ingress RevDNAT upwards

In a subsequent patch we want to call this code from
tail_nodeport_nat_ingress_ipv6(). Make it available without adding a
forward declaration.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: shuffle IPv4 Ingress RevDNAT upwards

In a subsequent patch we want to call this code from
tail_nodeport_nat_ingress_ipv4(). Make it available without adding a
forward declaration.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: extract return logic from rev_nodeport_lb*()

When rev_nodeport_lb*() doesn't result in a redirect to the egress path,
it circles back to the program's main path (or drops the packet, when
called from bpf_lxc).

Move the relevant logic into the caller, to make sharing it easier in the
future.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: wire up trace_ctx in Ingress RevDNAT path

Enable rev_nodeport_lb*() to utilize the trace information from a
previous CT lookup. While in most cases this would get overruled by the
actual RevDNAT CT lookup, EgressGW reply traffic can benefit from it.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: make Ingress RevDNAT tail-call conditional

Currently the nodeport Ingress path is split into two tail-calls for
program-size reasons. We first handle RevSNAT, then RevDNAT.

But we can take advantage of HAVE_LARGE_INSN_LIMIT on newer kernels, and
inline the RevDNAT handling into the RevSNAT tail-call. This enables us to
pass the trace_ctx through for better tracing.

While at it rename the RevDNAT functions to be more consistent with our
naming scheme.

Note that CILIUM_CALL_IPV*_NODEPORT_NAT_INGRESS is never invoked from
bpf_lxc, so we can remove the #ifdef's around edt_set_aggregate().

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nat: wire up trace_ctx for CT lookup in RevSNAT path

By returning the trace information to the caller, it can be passed onwards
to the Ingress RevDNAT path. There EgressGW reply traffic benefits from
the CT-driven trace aggregation.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

aspsk commented on bpf/lib/nodeport.h:
```diff
@@ -2452,16 +2451,17 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __s8 *ext_er
-		.reason = TRACE_REASON_UNKNOWN,
-		.monitor = 0,
-	};
+
```
Extra line

julianwiedmann replied:
nice catch, thanks! Fixed.

aspsk commented on bpf/lib/nodeport.h:
```diff
@@ -2544,15 +2567,29 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
+			goto recircle;
+
+		/* Redirected to egress interface: */
+		edt_set_aggregate(ctx, 0);
```
previously we set it only when !LXC. Can this be called from LXC?

julianwiedmann replied:
I added the following to the patch description, thank you!

>    Note that CILIUM_CALL_IPV*_NODEPORT_NAT_INGRESS is never invoked from  bpf_lxc, so we can remove the #ifdef's around edt_set_aggregate().

aspsk reviewed (commented):
LGTM, left one question and one cosmetic comment

aspsk reviewed (approved):
LGTM, thanks!

julianwiedmann committed:
bpf: nat: NULL-initialize the NAT entry in snat_v6_rev_nat()

e9876982b009 ("bpf: nat: share rewrite logic in RevSNAT path") already
added the same initialization in the IPv4 path. Do the same for IPv6.

This works around a verifier reject in one of the subsequent patches.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
